app/src/main/java/io/appium/settings/RecorderThread.java

```diff
@@ -0,0 +1,419 @@
+
+        /* Clamp raw width/height of device screen with video encoder's capabilities
+        *  to avoid crash.
+        * */
+        MediaCodecInfo.VideoCapabilities videoEncoderCapabilities = videoEncoder.getCodecInfo().
+                getCapabilitiesForType(videoMime).getVideoCapabilities();
+        this.videoWidth = videoEncoderCapabilities.getSupportedWidths().clamp(this.videoWidth);
+        this.videoHeight = videoEncoderCapabilities.getSupportedHeights().clamp(this.videoHeight);
+        this.videoBitrate = calcBitRate(videoWidth, videoHeight);
+
+        MediaFormat encoderFormat = MediaFormat.createVideoFormat(videoMime, videoWidth,
+                videoHeight);
+        encoderFormat.setInteger(MediaFormat.KEY_COLOR_FORMAT,
+                MediaCodecInfo.CodecCapabilities.COLOR_FormatSurface);
+        encoderFormat.setInteger(MediaFormat.KEY_BIT_RATE, videoBitrate);
+        encoderFormat.setInteger(MediaFormat.KEY_FRAME_RATE,
+                RecorderConstant.VIDEO_CODEC_FRAME_RATE);
+        encoderFormat.setInteger(MediaFormat.KEY_REPEAT_PREVIOUS_FRAME_AFTER,
+                RecorderConstant.AUDIO_CODEC_REPEAT_PREV_FRAME_AFTER);
+        encoderFormat.setInteger(MediaFormat.KEY_I_FRAME_INTERVAL,
+                RecorderConstant.AUDIO_CODEC_I_FRAME_INTERVAL);
+
+        videoEncoder.configure(encoderFormat, null, null,
+                MediaCodec.CONFIGURE_FLAG_ENCODE);
+        surface = videoEncoder.createInputSurface();
+        videoEncoder.start();
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    private void setupVirtualDisplay() {
+        virtualDisplay = mediaProjection.createVirtualDisplay("Appium Screen Recorder",
+                videoWidth, videoHeight, DisplayMetrics.DENSITY_HIGH,
+                DisplayManager.VIRTUAL_DISPLAY_FLAG_AUTO_MIRROR,
+                surface, displayCallback, handler);
+    }
+
+    private void setupAudioCodec() throws IOException {
+        // Encoded video resolution matches virtual display. TODO set channelCount 2 try stereo quality
+        MediaFormat encoderFormat = MediaFormat.createAudioFormat(MediaFormat.MIMETYPE_AUDIO_AAC,
+                sampleRate, RecorderConstant.AUDIO_CODEC_CHANNEL_COUNT);
+        encoderFormat.setInteger(MediaFormat.KEY_BIT_RATE,
+                RecorderConstant.AUDIO_CODEC_DEFAULT_BITRATE);
+
+        audioEncoder = MediaCodec.createEncoderByType(MediaFormat.MIMETYPE_AUDIO_AAC);
+        audioEncoder.configure(encoderFormat, null, null,
+                MediaCodec.CONFIGURE_FLAG_ENCODE);
+        audioEncoder.start();
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.Q)
+    private void setupAudioRecord() {
+        int channelConfig = AudioFormat.CHANNEL_IN_MONO;
+        int minBufferSize = AudioRecord.getMinBufferSize(sampleRate, channelConfig,
+                AudioFormat.ENCODING_PCM_16BIT);
+
+        AudioFormat.Builder audioFormatBuilder = new AudioFormat.Builder();
```

Comment:
I assume this intermediate variable is redundant. We could build a new audio format using the following syntax:

```
AudioFormat newAudioFormat = new AudioFormat.Builder()
  .setEncoding()
  ....
```

Reply:
same be

Reply:
done good catch thank you 👍